src/webserver.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the GunicornWebserver class to represent the gunicorn server."""
+import logging
+import pathlib
+import signal
+import typing
+
+from ops.model import Container
+from ops.pebble import ExecError, PathError
+
+from charm_state import CharmState
+from charm_types import ExecResult
+from consts import FLASK_SERVICE_NAME
+from exceptions import WebserverConfigInvalidError
+
+logger = logging.getLogger(__name__)
+
+
+class GunicornWebserver:
+    """A class representing a Gunicorn web server."""
+
+    def __init__(self, charm_state: CharmState, flask_container: Container):
+        """Initialize a new instance of the GunicornWebserver class.
+
+        Args:
+            charm_state: The state of the charm that the GunicornWebserver instance belongs to.
+            flask_container: The Flask container in this charm unit.
+        """
+        self._charm_state = charm_state
+        self._flask_container = flask_container
+
+    @property
+    def _config(self) -> str:
+        """Generate the content of the Gunicorn configuration file based on charm states.
+
+        Returns:
+            The content of the Gunicorn configuration file.
+        """
+        config_entries = [
+            f"bind = ['0.0.0.0:{self._charm_state.flask_port}']",
+            f"chdir = {repr(str(self._charm_state.flask_dir.absolute()))}",
+        ]
+        settings = ("workers", "threads", "keepalive", "timeout")
+        for setting in settings:
+            setting_value = getattr(self._charm_state.webserver_config, setting)
+            if setting_value is None:
+                continue
+            setting_value = (
+                setting_value
+                if isinstance(setting_value, int)
+                else int(setting_value.total_seconds())
+            )
+            config_entries.append(f"{setting} = {setting_value}")
+        return "\n".join(config_entries)
```

Comment:
Do you think this should be done using jinja? It would make it easier to see what the final configuration looks like?

Reply:
I found it difficult to use jinja to generate the configuration file since most of the configuration file is dynamic with very little static content. This makes it not very suited for template.

Reply:
Would it be possible to provide an example of what the jinja template would look like in this case please? It would be good to give a few people the chance to see it because based on the code, it is hard to visualise what it would look like right now

Reply:
Here's a simple example using jinja2:
```python
from jinja2.sandbox import SandboxedEnvironment
import pathlib
import datetime

template_content = """\
bind = ['0.0.0.0:{{ flask_port }}']
chdir = {{ repr(str(flask_dir.absolute())) }}
{%- for setting in ("workers", "threads", "keepalive", "timeout") -%}
    {%- if webserver_config.get(setting) %}
{{ setting }} = {{ render_setting_value(webserver_config.get(setting)) }}
    {%- endif %}
{%- endfor -%}
"""

env = SandboxedEnvironment()
env.globals["repr"] = repr
env.globals["str"] = str
template = env.from_string(template_content)

print(
    template.render(
        flask_port=8000,
        flask_dir=pathlib.Path("/srv/flask/app"),
        render_setting_value=lambda v: v if isinstance(v, int) else v.total_seconds(),
        webserver_config={"timeout": datetime.timedelta(seconds=1), "workers": 10},
    )
)
```

Reply:
I kind of prefer this, I can tell what the final file will look like much more easily. Any thoughts @gregory-schiano?

Reply:
What is the reason for needing `repr` and `str` around `flask_dir.absolute()`? I guess `flask_dir=repr(str(pathlib.Path("/srv/flask/app")))` could work as well? That way `env.globals` doesn't need to have `str` and `repr`?

Reply:
I think the problem of this jinja2 template is that almost everything is enclosed inside the `{% %}` and `{{ }}`, what make it less legible.

Reply:
The UI developer in me is used to that 🙂

Reply:
it is very common to have that in HTML rendering and other UI libraries

Reply:
> What is the reason for needing repr and str around flask_dir.absolute()? I guess flask_dir=repr(str(pathlib.Path("/srv/flask/app"))) in template.render could work as well? That way env.globals doesn't need to have str and repr?

I think format and rendering of the content is the responsibility of template, so I put the formatting inside the template.

Reply:
I don't feel like using a template engine for a simple `key = value` file worth it.
Templating is always risky in a project as it could lead to some mixed-ugly things, no template will avoid this.

But I'm wondering if the code could be simplified by considering every value as a string and preparing its value representations before and merge with the additional parameters we put (bind and chdir)

Reply:
maybe instead of using a templating engine we can use a nicely formatted multi-line f-string that looks much like the template for jinja above?

Reply:
I'm in :)

Reply:
To me that looks good, at least the structure is a bit clearer

Reply:
Okay, updated, thanks